diff --git a/src/app/layout.tsx b/src/app/layout.tsx
index 8b229bd..92586bf 100644
--- a/src/app/layout.tsx
+++ b/src/app/layout.tsx
@@ -1,7 +1,7 @@
 import { Analytics } from '@vercel/analytics/react';
 import { SpeedInsights } from '@vercel/speed-insights/react';
 import { clsx } from 'clsx/lite';
-import { IBM_Plex_Mono } from 'next/font/google';
+import localFont from 'next/font/local';
 import {
   BASE_URL,
   DEFAULT_THEME,
@@ -23,12 +23,28 @@ import AdminBatchEditPanel from '@/admin/AdminBatchEditPanel';
 import '../site/globals.css';
 import '../site/sonner.css';
 
-const ibmPlexMono = IBM_Plex_Mono({
-  subsets: ['latin'],
-  weight: ['400', '500', '700'],
+const ibmPlexMono = localFont({
+  src: [
+    {
+      path: './IBMPlexMono-Regular.ttf',
+      weight: '400',
+      style: 'normal',
+    },
+    {
+      path: './IBMPlexMono-Medium.ttf',
+      weight: '500',
+      style: 'normal',
+    },
+    {
+      path: './IBMPlexMono-Bold.ttf',
+      weight: '700',
+      style: 'normal',
+    },
+  ],
   variable: '--font-ibm-plex-mono',
 });
 
+
 export const metadata: Metadata = {
   title: SITE_TITLE,
   description: SITE_DESCRIPTION,
